src/js/protocols/webusbdfu.js

```diff
@@ -0,0 +1,1119 @@
+
+        this.chipInfo = null; // information about chip's memory
+        this.flash_layout = { 'start_address': 0, 'total_size': 0, 'sectors': [] };
+        this.transferSize = 2048; // Default USB DFU transfer size for F3,F4 and F7
+    }
+    connect(hex, options, callback) {
+        this.hex = hex;
+        this.callback = callback;
+
+        this.options = {
+            erase_chip: false,
+            exitDfu: false,
+        };
+
+        if (options.exitDfu) {
+            this.options.exitDfu = true;
+        } else if (options.erase_chip) {
+            this.options.erase_chip = true;
+        }
+
+        // reset and set some variables before we start
+        this.upload_time_start = new Date().getTime();
+        this.verify_hex = [];
+
+        // reset progress bar to initial state
+        TABS.firmware_flasher.flashingMessage(null, TABS.firmware_flasher.FLASH_MESSAGE_TYPES.NEUTRAL).flashProgress(0);
+
+        navigator.usb
+        .requestDevice(usbDevices)
+        .then(selectedDevice => {
+            console.log(`Product name: ${selectedDevice.productName}`);
+            console.log(`USB DFU detected with ID: ${selectedDevice}`);
+            this.usbDevice = selectedDevice;
+            console.log(`WebUSB Version: ${this.usbDevice.deviceVersionMajor}.${this.usbDevice.deviceVersionMinor}.${this.usbDevice.deviceVersionSubminor}`);
+            return this.openDevice();
+        })
+        .catch((e) => {
+            console.error(`There is no device. ${e}`);
+        });
+    }
+    openDevice() {
+        this.usbDevice
+        .open()
+        .then(async () => {
+            // show key values for the device
+            console.log(`USB Device opened: ${this.usbDevice.productName}`);
+            if (this.usbDevice.configuration === null) await this.usbDevice.selectConfiguration(1);
+            this.claimInterface(0);
+        })
+        .catch(error => {
+            console.log('Failed to open USB device:', error);
+            gui_log(i18n.getMessage('usbDeviceOpenFail'));
+        });
+    }
+    closeDevice() {
+        this.usbDevice
+        .close()
+        .then(() => {
+            gui_log(i18n.getMessage('usbDeviceClosed'));
+            console.log('DFU Device closed');
+        })
+        .catch(error => {
+            console.log('Failed to close USB device:', error);
+            gui_log(i18n.getMessage('usbDeviceCloseFail'));
+        });
+        this.usbDevice = null;
+    }
+    claimInterface(interfaceNumber) {
+        const self = this;
+
+        self.usbDevice
+        .claimInterface(interfaceNumber)
+        .then(() => {
+            console.log(`Claimed interface: ${interfaceNumber}`);
+
+            if (self.options.exitDfu) {
+                self.leave();
+            } else {
+                self.upload_procedure(0);
+            }
+        })
+        .catch(error => {
+            console.log('Failed to claim USB device', error);
+            self.cleanup();
+        });
```

Comment:
Seems like a leftover from old code.

We don't really need `self` pattern when using arrow functions and classes because `this` context is usually bound correctly.
```suggestion
        this.usbDevice
        .claimInterface(interfaceNumber)
        .then(() => {
            console.log(`Claimed interface: ${interfaceNumber}`);

            if (self.options.exitDfu) {
                this.leave();
            } else {
                this.upload_procedure(0);
            }
        })
        .catch(error => {
            console.log('Failed to claim USB device', error);
            this.cleanup();
        });
```

Reply:
Done. Line 145 should be `if (this.options.exitDfu)`